ui/src/main/java/edu/wpi/grip/ui/util/deployment/HeadlessInstanceManager.java

```diff
@@ -0,0 +1,155 @@
+package edu.wpi.grip.ui.util.deployment;
+
+
+import com.google.common.eventbus.EventBus;
+import com.jcabi.ssh.Shell;
+import com.jcraft.jsch.JSch;
+import edu.wpi.grip.core.StartStoppable;
+import edu.wpi.grip.core.events.StartedStoppedEvent;
+import edu.wpi.grip.core.events.UnexpectedThrowableEvent;
+import org.apache.tools.ant.BuildException;
+import org.apache.tools.ant.Project;
+import org.apache.tools.ant.taskdefs.optional.ssh.Scp;
+
+import java.io.File;
+import java.io.IOException;
+import java.net.URLDecoder;
+import java.nio.file.Paths;
+import java.util.Optional;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+
+/**
+ * Controls an instance of GRIP running on a remote device.
+ */
+public class HeadlessInstanceManager implements StartStoppable {
```

Comment:
This class never runs in headless mode afaik, so a better name might be "DeploymentManager"

8/10